spec/mem/tree_borrows/tree.md

```diff
@@ -130,5 +161,83 @@ impl Node {
         })
     }
 
+    /// Get a child node of `self`
+    /// `path` is the path from `self` to the target child node.
+    fn get_node(&mut self, path: Path) -> Result<Node> {
+        // `sub_root_id` indicates which node to access next; call it the sub-root.
+        // `sub_path` is the path from the sub-root to the target node.
+        let Some((sub_root_id, sub_path)) = path.split_first() else {
+            return ret(*self);
+        };
+
+        // If `self` is a leaf and the path keeps going, then the path is invalid.
+        if self.is_leaf() { panic!("Node::get_node: invalid node path"); }
+
+        // Find the right child, and recursively search for the target node.
+        self.children[sub_root_id].get_node(sub_path)
+    }
+
+    /// Check whether `self` is a leaf.
+    fn is_leaf(&self) -> bool {
+        self.children.len() == Int::ZERO
+    }
+}
+```
+
+In addition, we implement some methods for protector-related semantics.
+```rust
+impl Node {
+    /// Release the protector, and perform a special access for the protector end semantics.
+    /// Recusively do state transition on all foreigns of the protected node.
+    /// `path` is the path from `self` to the proctected node.
+    /// `path` is None when the protected node is not a descendant of `self`.
+    /// `location_states` are the location states of the protected node.
+    fn release_protector(
+        &mut self,
+        path: Option<Path>, // self -> protected node
+        location_states: &List<LocationState>,
+    ) -> Result {
+        let node_relation = if path.is_some() { NodeRelation::Child } else { NodeRelation::Foreign };
+
+        // If `self` is the protected node, we are done.
+        if path.is_some_and(|p| p.is_empty()) {
+            self.protected = Protected::No;
+            return ret(());
+        }
+
+        for offset in Int::ZERO..location_states.len() {
+            let LocationState { accessed, permission } = location_states[offset];
+
+            // If the location has never been accessed, there is no need to perform an access here.
+            if accessed != Accessed::Yes { continue; }
+
+            // If there was a write to the location (i.e. if the permission is Active),
+            // we perform a write access here. Otherwise, we perform a read access here.
```

Comment:
@JoJoDeveloping What is there was a write and then a foreign read? Then the permission will be frozen, but there was a write.  So the comment saying that this checks whether there was a write to the location doesn't seem right.

Reply:
That foreign read should be UB, I think
```rust
Permission::Active if protected.active(accessed) =>
                throw_ub!("Tree Borrows: reading from the foreign of an actively protected pointer with Active permission")
```

Reply:
Yes, this protector end semantics only happens with protectors which means that the foreign <anything> would have been UB. I.e. that can not happen.

Reply:
Ah, good point.

Please clarify the comment to say that.